src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+    fn upgrade_vocabulary(&mut self, definition: &Definition, from_version: Vocabulary) -> Result<VocabularyOutcome>;
+}
+
+impl Vocabulary {
+    // TODO: don't do linear search!
+    fn find<T>(&self, entid: T) -> Option<&Attribute> where T: Into<Entid> {
+        let to_find = entid.into();
+        self.attributes.iter().find(|&&(e, _)| e == to_find).map(|&(_, ref a)| a)
+    }
+}
+
+impl<'a, 'c> VersionedStore for InProgress<'a, 'c> {
+    fn verify_core_schema(&self) -> Result<()> {
+        if let Some(core) = self.read_vocabulary_named(&DB_SCHEMA_CORE)? {
+            if core.version != CORE_SCHEMA_VERSION {
+                bail!(ErrorKind::UnexpectedCoreSchema(Some(core.version)));
+            }
+
+            // TODO: check things other than the version.
+        } else {
+            // This would be seriously messed up.
+            bail!(ErrorKind::UnexpectedCoreSchema(None));
+        }
+        Ok(())
+    }
+
+    fn check_vocabulary<'definition>(&self, definition: &'definition Definition) -> Result<VocabularyCheck<'definition>> {
+        if let Some(vocabulary) = self.read_vocabulary_named(&definition.name)? {
+            // The name is present.
+            // Check the version.
+            if vocabulary.version == definition.version {
+                // Same version. Check that all of our attributes are present.
+                let mut missing: Vec<&'definition (NamespacedKeyword, Attribute)> = vec![];
+                for pair in definition.attributes.iter() {
+                    if let Some(entid) = self.get_entid(&pair.0) {
+                        if let Some(existing) = vocabulary.find(entid) {
+                            if *existing == pair.1 {
+                                // Same. Phew.
+                                continue;
+                            } else {
+                                // We have two vocabularies with the same name, same version, and
+                                // different definitions for an attribute. That's a coding error.
+                                // We can't accept this vocabulary.
+                                bail!(ErrorKind::ConflictingAttributeDefinitions(
+                                          definition.name.to_string(),
+                                          definition.version,
+                                          pair.0.to_string(),
+                                          existing.clone(),
+                                          pair.1.clone())
+                                );
+                            }
+                        }
+                    }
+                    // It's missing. Collect it.
+                    missing.push(pair);
+                }
+                if missing.is_empty() {
+                    Ok(VocabularyCheck::Present)
+                } else {
+                    Ok(VocabularyCheck::PresentButMissing { attributes: missing })
```

Comment:
`PresentButMissingAttributes`?  I read this as missing (the vocabulary) entirely, not as missing some attributes.

Reply:
Hm, I suppose I was feeling Swifty when I wrote that. I'll clarify.